### PR TITLE
Get severity outputs

### PR DESCRIPTION
**Pull request by pabloperguz**

### CODE REVIEW

File: R/combined.R [hunk 1]
```diff
@@ -286,17 +293,40 @@ combined_aggregate_data <- function(data) {
 }
 
 
-combined_aggregate_rt <- function(rt, samples) {
+aggregate_severity <- function(severity, samples) {
```
Comment (edknock) [inline R/combined.R:296]: `combined_aggregate_severity` would be more consistent

File: R/combined.R [hunk 2]
```diff
@@ -286,17 +293,40 @@ combined_aggregate_data <- function(data) {
...
+                                      weight = "admitted_inc")
+
+  infections <- combined_aggregate_rt(infections, samples,
+                                      weight = "infections_inc")
+
+  severity <- list(
+    admissions = admissions,
+    infections = infections
+  )
+}
```
Comment (edknock) [inline R/combined.R:316]: I think we want this to work similarly to combined_aggregate_rt - it takes the severity object as it is currently and adds in severity$england and severity$uk (if we have nations included), then outputs that

Comment (pabloperguz) [reply R/combined.R:316]: problem is we have different weights there

File: R/combined.R (hunk 1 above)
Comment (pabloperguz) [reply R/combined.R]: changed

File: R/fit_process.R [hunk 3]
```diff
@@ -51,6 +51,16 @@ spim_fit_process <- function(samples, parameters, data,
   message("Extracting demography")
   model_demography <- extract_demography(samples)
 
+  ## Check if severity calculations were outputted and, if so, extract
+  traj_names <- rownames(samples$trajectories$state)
+  if (all(c("ifr", "ihr", "hfr") %in% traj_names)) {
```
Comment (edknock) [inline R/fit_process.R:56]: better to do this with a switch I think - we also want to use the switch in lancelot_index

File: R/combined.R (hunk 2 above)
Comment (pabloperguz) [reply R/combined.R:316]: more so, as we add severity disaggregated arrays (e.g. by variant, etc) both objects will get bigger with the ones weighted by admissions and by infections, respectively

Comment (edknock) [reply R/combined.R:316]: It only makes sense to weight IFR and IHR by infections, and HFR by admissions

File: R/fit_process.R (hunk 3 above)
Comment (pabloperguz) [reply R/fit_process.R]: done; I think I'm having the same trouble as Yasin to use lancelot_index switches, so I've set it to TRUE as default in sircovid

File: R/combined.R [hunk 4]
```diff
@@ -286,17 +293,40 @@ combined_aggregate_data <- function(data) {
...
+  severity <- switch_levels(severity)
+  hosp_names <- grep("hfr", names(severity), value = TRUE)
+  inf_names <- grep("hfr", names(severity), invert = TRUE, value = TRUE)
+
+  admissions <- switch_levels(severity[c("date", "step", hosp_names)])
+  infections <- switch_levels(severity[inf_names])
+
+
+  admissions <- combined_aggregate_rt(admissions, samples,
+                                      weight = "admitted_inc")
```
Comment (edknock) [inline R/combined.R:307]: `admitted_inc` is just confirmed admissions - really you'd want `admitted_inc + diagnoses_inc`

Comment (pabloperguz) [reply R/combined.R]: This is now taken care of in the refactored function

File: R/combined.R [hunk 5]
```diff
@@ -286,6 +297,55 @@ combined_aggregate_data <- function(data) {
...
+combined_aggregate_severity_1 <- function(x, samples, weight) {
+  england <- sircovid::regions("england")
+  nations <- sircovid::regions("nations")
+
+  if (all(england %in% names(x))) {
+    x$england <- sircovid::combine_rt(x[england], samples[england],
+                                       rank = FALSE, weight = weight)
+  }
+  if (all(nations %in% names(x))) {
+    stop("We cannot currently run severity analysis at DVN/UK level!")
```
Comment (edknock) [inline R/combined.R:342]: Why do we need this?

File: R/combined.R [hunk 6]
```diff
@@ -286,6 +297,55 @@ combined_aggregate_data <- function(data) {
...
 
 
+combined_aggregate_severity <- function(severity, samples) {
+
+  # We first need to sum admitted_inc and diagnoses_inc trajectories to get
+  # all_admissions_inc, which we'll use as weight to aggregate hospital severity
+  for (i in names(samples)) {
+    admitted_inc <- samples[[i]]$trajectories$state["admitted_inc", , ]
+    diagnoses_inc <- samples[[i]]$trajectories$state["diagnoses_inc", , ]
+    all_admissions <- array(admitted_inc + diagnoses_inc, dim = c(1, 10, 744))
```
Comment (edknock) [inline R/combined.R:307]: I assume you don't want the dim hardcoded like this?

Comment (edknock) [reply R/combined.R:307]: Probably best to add `drop = FALSE` in the above two lines and then can just have `all_admissions <- admitted_inc + diagnoses_inc`

File: R/fit_process.R [hunk 7]
```diff
@@ -51,6 +54,15 @@ spim_fit_process <- function(samples, parameters, data,
...
   model_demography <- extract_demography(samples)
 
+  ## Check get_severity is TRUE extract
+  if (get_severity) {
+    message("Extracting severity outputs")
+    severity <- extract_severity(samples$trajectories$state)
+    # Add Rt elements so severity has the same structure for post-processing
+    severity[["step"]] <- rt$step
+    severity[["date"]] <- rt$date
+  } else (severity <- NULL)
```
Comment (edknock) [inline R/fit_process.R:64]: curly brackets here please

File: R/combined.R (hunk 5 above)
Comment (pabloperguz) [reply R/combined.R:342]: changed

File: R/combined.R (hunk 6 above)
Comment (pabloperguz) [reply R/combined.R:307]: makes sense... thanks!

File: R/fit_process.R (hunk 7 above)
Comment (pabloperguz) [reply R/fit_process.R:64]: changes

File: R/combined.R (hunk 5 above)
Comment (edknock) [reply R/combined.R]: Why not aggregate to UK level though - I don't see why we can't here

File: R/control.R [hunk 8]
```diff
@@ -92,10 +95,19 @@ spim_control <- function(short_run, n_chains, deterministic = FALSE,
...
+    severity <- FALSE
+  }
+
   particle_filter <- list(n_particles = n_particles,
                           n_threads = n_threads,
                           seed = NULL,
-                          compiled_compare = compiled_compare)
+                          compiled_compare = compiled_compare,
+                          severity = severity,
+                          multiregion = multiregion)
```
Comment (edknock) [inline R/control.R:110]: `multiregion` not needed here with change suggested https://github.com/mrc-ide/spimalot/pull/208/files#r857780215

**Review by edknock — APPROVED** [review]

**Pull request merged**